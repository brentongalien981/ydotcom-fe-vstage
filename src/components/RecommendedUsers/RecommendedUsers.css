.user-list-item {
  margin-bottom: 10px;
}

.user-list .profile-image {
  width: 50px;
  border: 1px solid #CCC;
}

.profile-link {
  text-decoration: none;
  margin: 0 5px;
}