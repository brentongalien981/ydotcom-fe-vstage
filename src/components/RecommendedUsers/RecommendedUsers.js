import React, { useEffect } from "react";
import { Button, Image, Spinner } from "react-bootstrap";
import "./RecommendedUsers.css";
import { Link } from "react-router-dom";
import { useDispatch, useSelector } from "react-redux";
import { selectorError, selectorIsReading, selectorRecommendedUsers } from "../../redux/selectors/recommendedUsersSelector";
import { readRecommendedUsers } from "../../redux/actions/recommendedUsersActions";


const RecommendedUsers = () => {

  // Use hooks.
  const dispatch = useDispatch();

  // Reference redux state.
  const recommendedUsers = useSelector(selectorRecommendedUsers);
  const isReading = useSelector(selectorIsReading);
  const error = useSelector(selectorError);


  // Do initial reading of recommended users.
  useEffect(() => {
    dispatch(readRecommendedUsers());
  }, []);



  // Set the component.
  const list = recommendedUsers.map((user) => {

    const profilePhotoSource = user.Profile?.photoSource ? `/photos/${user.Profile.photoSource}` : `/photos/penguin.jpg`;
    const profileLink = `/profile/${user.username}`;
    const displayedUsername = `@${user.username.substring(0, 12)}`;

    return (
      <div className="user-list-item d-flex align-items-center" key={user.username}>
        <Image className="profile-image" src={profilePhotoSource} roundedCircle />
        <Link to={profileLink} className="profile-link">{displayedUsername}</Link>
        <Button size="sm" className="ms-auto">Follow</Button>
      </div>
    );
  });


  // Set loader component.
  let loaderComponent = null;
  if (isReading) {
    loaderComponent = (
      <div className="d-flex justify-content-center" style={{ padding: "20px" }}>
        <Spinner animation="border" variant="primary" className="animate__animated animate__fadeIn" data-testid="recommendedUsersSpinner" />
      </div>
    );
  }


  // Set error component.
  let errorComponent = null;
  if (error) {
    errorComponent = (<p>{error}</p>);
  }


  // Return the component.
  return (
    <div className="user-list" data-testid="recommendedUsersTestId">
      {list}
      {loaderComponent}
      {errorComponent}
    </div>
  )

};


export default RecommendedUsers;