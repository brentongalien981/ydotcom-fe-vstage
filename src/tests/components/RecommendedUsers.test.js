import { act, screen } from "@testing-library/react";
import RecommendedUsers from "../../components/RecommendedUsers/RecommendedUsers";
import { renderWithProviders } from "../../utils/test-utils/renderWithProviders";

it("renders initial <RecommendedUsers /> correctly", async () => {

  // Render with redux and the store.
  let container;
  act(() => {
    const { asFragment } = renderWithProviders(
      <RecommendedUsers />
    );
    container = asFragment;
  });

  // Reference the component.
  const recommendedUsersComponent = screen.getByTestId("recommendedUsersTestId");

  // Assert that the component is rendered
  expect(recommendedUsersComponent).toBeInTheDocument();

  // Assert that the snapshot matches.
  expect(container()).toMatchSnapshot();

});