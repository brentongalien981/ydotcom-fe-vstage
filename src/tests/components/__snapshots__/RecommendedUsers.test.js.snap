// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders initial <RecommendedUsers /> correctly 1`] = `
<DocumentFragment>
  <div
    class="user-list"
    data-testid="recommendedUsersTestId"
  >
    <div
      class="d-flex justify-content-center"
      style="padding: 20px;"
    >
      <div
        class="animate__animated animate__fadeIn spinner-border text-primary"
        data-testid="recommendedUsersSpinner"
      />
    </div>
  </div>
</DocumentFragment>
`;
